keystore/keystore.go

```diff
@@ -0,0 +1,222 @@
+package keystore
+
+import (
+	"context"
+	"crypto/ed25519"
+	"encoding/json"
+	"fmt"
+	"sync"
+	"time"
+
+	"golang.org/x/crypto/curve25519"
+
+	gethkeystore "github.com/ethereum/go-ethereum/accounts/keystore"
+	gethcrypto "github.com/ethereum/go-ethereum/crypto"
+	"github.com/smartcontractkit/chainlink-common/keystore/internal"
+	"github.com/smartcontractkit/chainlink-common/keystore/serialization"
+	"github.com/smartcontractkit/chainlink-common/keystore/storage"
+	"google.golang.org/protobuf/proto"
+)
+
+type KeyType string
+
+const (
+	X25519 KeyType = "X25519"
+	// TODO: Support P256 for DKG.
+	// Digital signature key types.
+	Ed25519        KeyType = "ed25519"
+	EcdsaSecp256k1 KeyType = "ecdsa-secp256k1"
+)
+
+var AllKeyTypes = []KeyType{X25519, Ed25519, EcdsaSecp256k1}
+
+type ScryptParams struct {
+	N int
+	P int
+}
+
+var (
+	DefaultScryptParams = ScryptParams{
+		N: gethkeystore.StandardScryptN,
+		P: gethkeystore.StandardScryptP,
+	}
+	FastScryptParams ScryptParams = ScryptParams{
+		N: 1 << 14,
+		P: 1,
+	}
+)
+
+// KeyInfo is the information about a key in the keystore.
+// Public key may be empty for non-asymmetric key types.
+type KeyInfo struct {
+	Name      string
+	KeyType   KeyType
+	CreatedAt time.Time
+	PublicKey []byte
+	Metadata  []byte
+}
+
+// newKeyInfo is a private constructor that ensures all fields are set explicitly.
+func newKeyInfo(name string, keyType KeyType, createdAt time.Time, publicKey []byte, metadata []byte) KeyInfo {
+	return KeyInfo{
+		Name:      name,
+		KeyType:   keyType,
+		CreatedAt: createdAt,
+		PublicKey: publicKey,
+		Metadata:  metadata,
+	}
+}
+
+type Keystore interface {
+	Admin
+	Reader
+	Signer
+	Encryptor
+}
```

Comment:
We should probably use the unexported method trick to be able to grow this interface in a compatible way:
```suggestion
type Keystore interface {
	Admin
	Reader
	Signer
	Encryptor
	
	mustEmbedUnimplementedKeystore()
}

type UnimplementedKeystore struct {}

func (UnimplementedKeystore) mustEmbedUnimplementedKeystore() {}
```

Reply:
(and perhaps the sub-interfaces? 🤷 )